### PR TITLE
On hard not OK to hard not OK change don't suppress (if needed) notification

### PR DESCRIPTION
**Pull request by Al2Klimov**
but discard it for the case of a recovery to be also suppressed (e.g. during downtime).

Before:

1. Checkable became hard critical
2. Problem notification was sent
3. Downtime started
4. Checkable became hard unknown (notification was suppressed)
5. Checkable became OK (notification was not suppressed,
   but discarted together with the unknown one)
6. Downtime ended
7. No recovery notification

After:

1. Checkable becomes hard critical
2. Problem notification is sent
3. Downtime starts
4. Checkable becomes hard unknown
   (notification isn't suppressed, but discarted)  <-- THIS PATCH
5. Checkable becomes OK (notification is suppressed)
6. Downtime ends
7. Recovery notification is (unsuppressed and) sent

refs #9054
closes #9146

### CODE REVIEW

**Review by julianbrost — CHANGES_REQUESTED** [review]
With the PR, the RECOVERY notification is still missing if the service temporarily recovers during the downtime but fails again, see the screenshot below. I think notifications should be sent if the state changed compared to the last state a notification was sent for. Or in other words: notifications should behave somewhat like as if nothing has happened within the downtime interval.

20211216_12h30m59s_grim

**Comment by Al2Klimov** [comment]
Other construction area: #9146

**Comment by lippserd** [comment]
Any objection to fixing everything in this PR instead of having them separated? I don't see any advantage in this and they have to be tested together anyway. For me it's not just another construction area.

**Comment by julianbrost** [comment]
@Al2Klimov Do the recent pushes now supersede #9146?

**Comment by Al2Klimov** [comment]
Yes, see my description edit.

**Review by julianbrost — CHANGES_REQUESTED** [review]
With this PR, another notification goes missing instead: if a service starts its downtime in a HARD WARNING state, worsens to HARD CRITICAL during the downtime, with the current master, a problem notification is sent after the downtime ends whereas with this PR, this notification no longer is sent. See the following screenshot (where I've sent the custom notification, I'd have expected a problem notification):



<details>
<summary>Test Script</summary>

```shell
#!/bin/bash

service='master-1!no-active-checks'
version=$(curl -ksSI https://localhost:5665/v1 | grep '^Server:' | cut -d: -f2 | tr -d '[:space:]')

# ensure hard ok as initial state
curl -iksSu root:icinga -H 'Accept: application/json' -X POST https://localhost:5665/v1/actions/process-check-result -d@- <<EOF
{
  "type": "Service",
  "service": "$service",
  "exit_status": 0,
  "plugin_output": "$version",
  "pretty": true
}
EOF
sleep 1

# bring service into hard warning state
for i in {1..5}; do
	curl -iksSu root:icinga -H 'Accept: application/json' -X POST https://localhost:5665/v1/actions/process-check-result -d@- <<EOF
{
  "type": "Service",
  "service": "$service",
  "exit_status": 1,
  "plugin_output": "$version",
  "pretty": true
}
EOF
	sleep 1
done


# schedule downtime
curl -ksSiu root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/actions/schedule-downtime' -d@- <<EOF
{
  "type": "Service",
  "service": "$service",
  "author": "test",
  "comment": "$version",
  "start_time": $(date +%s),
  "end_time": $(date -d '+1 hour' +%s),
  "fixed": true,
  "pretty": true
}
EOF
sleep 1

# bring service into hard warning state
curl -ksSu root:icinga -H 'Accept: application/json' -X POST https://localhost:5665/v1/actions/process-check-result -d@- <<EOF
{
  "type": "Service",
  "service": "$service",
  "exit_status": 2,
  "plugin_output": "$version",
  "pretty": true
}
EOF
sleep 1

# cancel downtime
curl -ksSiu root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/actions/remove-downtime' -d@- <<EOF
{
  "type": "Service",
  "service": "$service",
  "author": "test",
  "pretty": true
}
EOF
```
</details>

**Comment by Al2Klimov** [comment]
I'm sorry. I'm afraid that's how suppressed notifications work. For your case reminders fit best.

**Comment by julianbrost** [comment]
How is this how suppressed notifications work? I mean it worked before and this PR breaks it.

File: lib/notification/notificationcomponent.cpp [hunk 1]
```diff
@@ -82,9 +82,15 @@ void FireSuppressedNotifications(const Notification::Ptr& notification)
...
-		if ((suppressedTypes & type) && !checkable->NotificationReasonApplies(type)) {
-			subtract |= type;
-			suppressedTypes &= ~type;
+		if (suppressedTypes & type) {
+			switch (checkable->NotificationReasonApplies(type)) {
+				case NotifyReasonApplies::Yes:
+					break;
+				case NotifyReasonApplies::No:
+					subtract |= type;
+				default:
```
[the remark this answers is not in the document]
Comment (julianbrost) [reply lib/notification/notificationcomponent.cpp:91]: Well, yet another language feature that is deemed so error-prone that it's worth its own compiler warning (`-Wimplicit-fallthrough`).

**Comment by lippserd** [comment]
> How is this how suppressed notifications work? I mean it worked before and this PR breaks it.

I agree with that. In general, I would say that this PR must not influence behavior that is working properly at the moment.

File: lib/notification/notificationcomponent.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/notification/notificationcomponent.cpp:91]: C'mon. Really. We're not talking about goto and event that language feature may be used **to improve** code readability (just grep this repo). Kinda sweet from the compiler that it tries to protect me from myself, but no thanks.

**Comment by Al2Klimov** [comment]
IMAO we can break either what you mentioned or v2.1x.y v2.1x.z interoperability on suppressed notifications.

@julianbrost Please make a suggestion how to fix this.

**Comment by julianbrost** [comment]
So, what's the underlying issue here? You only have one bit available for whether a problem notification has to be sent but you need more information to differentiate between warning and critical?

The interoperability you mentioned should only affect the current zone as the suppression should only be needed there. I think it should be possible to implement a graceful fallback so that if you'd mix 2.13.2 and 2.13.3 in a zone, it behaves no worse than at the moment, but for it to be fixed, you'd have to upgrade both nodes.

An alternative I was thinking about would be to just always send a problem notification at the end of a downtime on non-OK states. Rationale: you did some maintenance work, probably assuming that everything was working when you were done. If it's not, there should always be a notification. At the moment, whether you'd get a notification depends on the state at downtime start. But that's probably more than what we'd want to change in a bugfix release and also needs some more consideration on the effects this would have (Notifications only for problem/recovery, not downtimes; how does it affect the use of notifications for automation, like creating support tickets, ...)

So if we want to get this in 2.13.3/2.12.7, trying to implement it with a graceful fallback is probably the way to go, otherwise we could also do a more invasive fix and only release it with 2.14.0. In general I'd say it's less bad if we lose a recovery notification than a problem notification, so I think the current behavior is more acceptable (if not ideal) than the current state of this PR.

**Comment by Al2Klimov** [comment]
> just always send a problem notification at the end of a downtime on non-OK states

> the current behavior is more acceptable (if not ideal) than the current state of this PR

Let's hear @lippserd's opinion...

File: lib/notification/notificationcomponent.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lippserd) [reply lib/notification/notificationcomponent.cpp:91]: At least it's worth commenting that this is on purpose.

**Comment by lippserd** [comment]
> > just always send a problem notification at the end of a downtime on non-OK states
> 
> > the current behavior is more acceptable (if not ideal) than the current state of this PR
> 
> Let's hear @lippserd's opinion...

I opt for fixing this bug 😆  I don't think it's worth even thinking about how this is going to work with mixed versions in zones. It's just not supported at all.

**Comment by julianbrost** [comment]
So without completely changing how suppressed notifications work, this probably requires not only keep track of a single bit for `NotificationProblem` in `suppressed_notification`, but rather one bit for `NotificationWarning` and one for `NotificationCritical` (at least for services, hosts only know up and down).

**Comment by Al2Klimov** [comment]
And in your opinion: during suppression, which upcoming `Notification$STATE`s would neutralise which other ones? Btw. you forgot unknown:

https://github.com/Icinga/icinga2/blob/463b15941401efd360502b37604a3e675e8ba27a/lib/icinga/checkresult.ti#L14-L31

> the current behavior is more acceptable (if not ideal)

Due to the complexity which already came up I also think in a such case recovery notifications (in contrast to problem ones) are dispensable.

**Comment by Al2Klimov** [comment]
@lippserd Didn't you suggest to just save the last notification while suppressed? This would be kinda easy and nothing would get lost. We'd just have not to let neutralise problem and recovery each other (problem suppressed + recovery suppressed = 0 suppressed). Instead they'd replace each other. However, then:

IMG_0979
IMG_0980

OK?

**Pull request closed without merging**